explainaboard/processors/ranking.py

```diff
@@ -0,0 +1,140 @@
+                    info, x['utterance'][int(x['true_label'])]
+                ),
+            ),
+            "num_oov": feature.Value(
+                dtype=feature.DataType.FLOAT,
+                description="the number of out-of-vocabulary words",
+                require_training_set=True,
+                func=lambda info, x, c, stat: feat_num_oov(
+                    info, x['context'], stat['vocab']
+                ),
+            ),
+            "fre_rank": feature.Value(
+                dtype=feature.DataType.FLOAT,
+                description=(
+                    "the average rank of each word based on its frequency in "
+                    "training set"
+                ),
+                require_training_set=True,
+                func=lambda info, x, c, stat: feat_freq_rank(
+                    info, x['context'], stat['vocab_rank']
+                ),
+            ),
+            "length_fre": feature.Value(
+                dtype=feature.DataType.FLOAT,
+                description="the frequency of context length in training set",
+                require_training_set=True,
+                func=lambda info, x, c, stat: feat_length_freq(
+                    info, x['context'], stat['length_fre']
+                ),
+            ),
+        }
+
+        return [
+            AnalysisLevel(
+                name='example',
+                features=features,
+                metric_configs=self.default_metrics(),
+            )
+        ]
+
+    def default_analyses(self) -> list[Analysis]:
+        """See Processor.default_analyses."""
+        analyses: list[Analysis] = []
+        analyses.extend(self.continuous_feature_analyses())
+        return analyses
+
+    @classmethod
+    def default_metrics(
+        cls, level='example', source_language=None, target_language=None
+    ) -> list[MetricConfig]:
+        """See Processor.default_metrics."""
+        return [AccuracyConfig(name='Accuracy')]
+
+    def _statistics_func(self, samples: Iterable[Any], sys_info: SysOutputInfo):
+        """See Processor._statistics_func."""
+        vocab: dict[str, float] = {}
+        length_fre: dict[int, float] = {}
+        total_samps = 0
+        tokenizer = unwrap(sys_info.source_tokenizer)
+        for sample in progress(samples):
+            text = sample["context"]
+            tokens = tokenizer(text)
+            length = len(tokens)
+
+            length_fre[length] = length_fre.get(length, 0.0) + 1.0
+
+            # update vocabulary
+            for w in tokens:
+                vocab[w] = vocab.get(w, 0.0) + 1.0
+
+            total_samps += 1
+
+        # the rank of each word based on its frequency
+        sorted_dict = {
+            key: rank
+            for rank, key in enumerate(sorted(set(vocab.values()), reverse=True), 1)
+        }
+        vocab_rank = {k: sorted_dict[v] for k, v in vocab.items()}
+
+        for k, v in length_fre.items():
+            length_fre[k] = v * 1.0 / total_samps
```

Comment:
- It isn't good to reuse the same variable for different meanings.
- Single slash represents division of floats. We basically don't need the magic constant `1.0`.
```suggestion
        # length_fre should be renamed to length_counts or something similar.
        length_freq = {k: v / total_samps for k, v in length_fre.items()}
```

Reply:
Maybe the L138 runs into zero-division error. Some workaround should be implemented.